### PR TITLE
Fix request sending on permissions change

### LINKED ISSUE #321: "Return to WEB_VIEW" button is not displayed in the Options on HMI
**Bug Report**
"Return to WEB_VIEW" button is not displayed in the Options on HMI
WebEngine App does not send AddCommand RPC ( Rpc: "AddCommand" Type: "request" Params: {"cmdID":1001,"menuParams":{"menuName":"Return to WEB_VIEW","position":0,"parentID":0}})) to SDL

**Preconditions:**

1. HMI and SDL are started
2. Web app is registered and activated successfully
HMI displays the WEB_VIEW template

**Reproduction Steps**

Press "Show menu" button on HMI
1. Web app sends ShowAppMenu: {}

**Expected Behavior**

Adding return to WEB_VIEW template command
1. Web app sends AddCommand request
AddCommand: {"cmdID":1001,"menuParams":{"menuName":"Return to WEB_VIEW","position":0,"parentID":0}}
2. SDL → HMI: UI.AddCommand: { "appID":753901101,"cmdID":1001,"menuParams":
{"menuName":"Return to WEB_VIEW","parentID":0,"position":0}}}
3. HMI -> SDL: UI.AddCommand:{ code":0 }
4. SDL → Web app: AddCommand:{"resultCode": "SUCCESS ", "success":true}
5. "Return to WEB_VIEW" button is displayed in Options on HMI*

**Observed Behavior**
"Return to WEB_VIEW" button is not displayed  in Options on HMI
Web app does not send AddCommand(Return to WEB_VIEW) request

OS & Version Information

OS/Version: Ubuntu 18.04
Sdl_javascript_suite: https://github.com/smartdevicelink/sdl_javascript_suite(3220edd)
HMI: https://github.com/smartdevicelink/sdl_hmi/tree/develop(6d32f1b)

Logs:
[SDL_HMI_logs.zip](https://github.com/smartdevicelink/sdl_javascript_suite/files/5260867/SDL_HMI_logs.zip)

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
Fixes #321 and #320 and #323 

This PR **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Will be tested manually

### Summary
Logic of sending AddCommand and SubscribeButton for sample WEB_VIEW application were moved from OnHmiStatus callback to permissions change callback. This is more reliable way to trigger sending of initial RPCs as changing of HMI level is not always guarantees that RPC is allowed. Also, this is a more simple way to execute some blocks of code just once.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) and this pull request adheres to the [Contributing Guide](./CONTRIBUTING.md).

### CODE REVIEW

**Comment by AKalinich-Luxoft** [comment]
@theresalech Please notice that this PR is ready for Livio review.

**Comment by crokita** [comment]
@VjKlepikov This PR seems to try to fix more than what the issues it references suggest. I see changes regarding how to subscribe to the buttons to resolve #323. I believe this change is fixing the wrong issue: I have posted a comment on #323 with further details, and what the changes should likely be instead. 

For changing the logic to use permission changes instead of the HMI status, the PR assumes that policies are used in Core, which isn't required. So, there's a scenario where that logic may never run, as opposed to listening for the OnHMIStatus. Additionally, I did not see how that change would help in resolving either of the issues this PR references: if it is meant to help resolve #323 then I believe the comment I gave on that issue is the likely solution. If there is an issue with the main app logic in `_onHmiStatusListener` being run more than once then I request creating an issue that describes how the problem can happen and a separate PR not requiring use of listening to permission changes.

I would be more comfortable approving this PR if it only resolves #321, as I agree with the changes and the issue being accurate.

**Comment by AKalinich-Luxoft** [comment]
> @VjKlepikov This PR seems to try to fix more than what the issues it references suggest. I see changes regarding how to subscribe to the buttons to resolve #323. I believe this change is fixing the wrong issue: I have posted a comment on #323 with further details, and what the changes should likely be instead.
> 
> For changing the logic to use permission changes instead of the HMI status, the PR assumes that policies are used in Core, which isn't required. So, there's a scenario where that logic may never run, as opposed to listening for the OnHMIStatus. Additionally, I did not see how that change would help in resolving either of the issues this PR references: if it is meant to help resolve #323 then I believe the comment I gave on that issue is the likely solution. If there is an issue with the main app logic in `_onHmiStatusListener` being run more than once then I request creating an issue that describes how the problem can happen and a separate PR not requiring use of listening to permission changes.
> 
> I would be more comfortable approving this PR if it only resolves #321, as I agree with the changes and the issue being accurate.

@crokita let me explain a few things here:
1. I agree with your point that code moved to policy listener may be never run, so I reverted these changes
2. Issue here is more complex than it looks so it make sense to fix it in one PR
3. First issue is that app is trying to subscribe to the same buttons each when it receives `OnHMIStatus`, even if it is already subscribed to it. For example, activate app, deactivate it, and activate again. Or, just click "Show menu" button as suggested in reproduction steps of #323. It can be fixed by adding a new flag. Similar changes was already done for initial AddCommand.
4. Second issue is that `screenManager.addButtonListener` throws the exception if SDL responds with any kind of error. So even if first issue is being fixed and app will make a subscription just once, due to unhandled exception app does not perform subscription to the rest of buttons.
5. Third issue has dependency on second. Because of exception thrown, app does not run the rest of the code, including the code related to sending of AddCommand (issue #321). So fixing of the second issue automatically solves the third issue. This is also a reason why these two issues can't be separated.
6. I also see that app is trying to subscribe to RC buttons, however it is not an RC application according to app manifest. That means that this app will never have a success with it. By that reason, RC buttons was replaced with the regular buttons. Please let me know if you agree with that?

PR has been simplified and a new changes were introduced in https://github.com/smartdevicelink/sdl_javascript_suite/pull/319/commits/c331dec93be0026ae956f70c75dba30c15d427cf

**Comment by crokita** [comment]
@AKalinich-Luxoft I see the issue regarding the app executing its main logic multiple times. The buttons get subscribed to multiple times when closing the application and re-activating it. I think the changes made to this PR make sense now. 

I think it's fine for point 6. for the WebEngine app to subscribe to those buttons because the app type is not remote control as you've mentioned. Furthermore, the issues found here will apply to the other example apps in this project, regarding over subscription of buttons and not catching for subscription errors. Can you apply the fixes to the `node/hello-sdl`, `node/hello-sdl-tcp` and `js/hello-sdl` example apps as well?

**Comment by AKalinich-Luxoft** [comment]
> @AKalinich-Luxoft I see the issue regarding the app executing its main logic multiple times. The buttons get subscribed to multiple times when closing the application and re-activating it. I think the changes made to this PR make sense now.
> 
> I think it's fine for point 6. for the WebEngine app to subscribe to those buttons because the app type is not remote control as you've mentioned. Furthermore, the issues found here will apply to the other example apps in this project, regarding over subscription of buttons and not catching for subscription errors. Can you apply the fixes to the `node/hello-sdl`, `node/hello-sdl-tcp` and `js/hello-sdl` example apps as well?

@crokita It's a good idea to apply the same changes to other example apps.
Done in https://github.com/smartdevicelink/sdl_javascript_suite/pull/319/commits/8e2747e2a8a851d500cfba1de3cf33a5a9ab299d
Please double check if it works as expected now.

**Review by crokita — CHANGES_REQUESTED** [review]
See feedback @AKalinich-Luxoft

File: examples/js/hello-sdl/index.html [hunk 1]
```diff
@@ -131,15 +132,23 @@
...
-                    const ButtonName = SDL.rpc.enums.ButtonName;
-                    const buttonNames = [ButtonName.AC_MAX, ButtonName.AC, ButtonName.RECIRCULATE, ButtonName.FAN_UP, ButtonName.FAN_DOWN, ButtonName.TEMP_UP,
-                        ButtonName.TEMP_DOWN, ButtonName.FAN_DOWN, ButtonName.DEFROST_MAX, ButtonName.DEFROST_REAR, ButtonName.DEFROST, ButtonName.UPPER_VENT,
-                        ButtonName.LOWER_VENT, ButtonName.VOLUME_UP, ButtonName.VOLUME_DOWN, ButtonName.EJECT, ButtonName.SOURCE, ButtonName.SHUFFLE, ButtonName.REPEAT];
-
-                    for (const buttonName of buttonNames) {
-                        await screenManager.addButtonListener(buttonName, this._onButtonListener.bind(this));
+                    if (!this._isButtonSubscriptionRequested) {
+                        // add button listeners
+                        const screenManager = this._sdlManager.getScreenManager();
```
[the remark this answers is not in the document]
Comment (AKalinich-Luxoft) [reply examples/js/hello-sdl/index.html]: @crokita good catch!
Last commit has been amended - https://github.com/smartdevicelink/sdl_javascript_suite/pull/319/commits/f8c40df5daf59b6fcc70eecd6e66eefb8b3f4b26

**Review by crokita — APPROVED** [review]

**Pull request merged**